torch/_jit_internal.py

```diff
@@ -31,12 +31,50 @@ def env(qualified_name, module):
         # or `a.b.c.d`. We first look up `torch` or `a` in the function's closed
         # over scope, then proceed to use the looked-up value to go down the
         # chain.
-        if '.' in qualified_name:
+        if '[' in qualified_name:
+            first, last = qualified_name.find('['), qualified_name.rfind(']')
+            base = qualified_name[:first]
+            base_mod = env(base, module)
+
+            # intentional bailout, e.g. base == "Tuple" but wasn't imported in Module
+            if not base_mod:
+                return None
+
+            # assume only subexp (between []) could contain ','
+            # but allow each element of the list to be any type hence recursive env()
+            subexp = qualified_name[first + 1: last]
+            if ',' in subexp and (subexp.find('[') == -1 or
+                                  subexp.find('[') > subexp.find(',')):
+                no_whitespace = "".join(subexp.split())
+                parts = no_whitespace.split(',')
+                types = [env(p, module) for p in parts]
+
+                # HACK for now - must be a better way to approximate base_mod[*types]
+                if len(types) == 1:
+                    return base_mod[types[0]]
+                elif len(types) == 2:
+                    return base_mod[types[0], types[1]]
+                elif len(types) == 3:
+                    return base_mod[types[0], types[1], types[2]]
+                elif len(types) == 4:
+                    return base_mod[types[0], types[1], types[2], types[3]]
+                else:
+                    raise ValueError("tuples beyond length 4 not supported yet.")
+
+            else:
+                # either base or subexp could include '.' or '['
+                subexp_mod = env(subexp, module)
+                return base_mod[subexp_mod]
+        elif '.' in qualified_name:
             parts = qualified_name.split('.')
             base = parts[0]
             remainding_pieces = '.'.join(parts[1:])
             module_value = getattr(module, base)
             return env(remainding_pieces, module_value)
+        elif 'int' == qualified_name:
```

Comment:
thoughts on doing this here? or should these type of things result in a bailout back to cpp too?

if this is the right approach, then i'd probably want to be more thorough about the types we handle.

Reply:
i'll come back to this one; i only added it to fix some failures in running test_jit.py, and in those cases getattr(module, 'int') was failing for me i thought.

Reply:
yea these were able to go away.  i had added them in to good effect, they made a test pass. but taking them out and adding in more thorough 'fallback' (return None) coverage fixed it too.